File: 실버 3 (3문제)/1463_1로 만들기.c
// 참고 링크 : https://moon-gd.tistory.com/49
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX 1000001

int numbers[MAX] = { 0, 0, 1, 1 };

int main() {
	int size = 0;
	scanf("%d", &size);

	// dp 계산
	for (int i = 4; i <= size; i++) {
		// 우선 i - 1 번째 값 + 1 을 저장
		numbers[i] = numbers[i - 1] + 1;

		// 2로 나누어지는 경우 확인
		if (i % 2 == 0 && numbers[i / 2]) {
			numbers[i] = numbers[i] < numbers[i / 2] + 1 ? numbers[i] : numbers[i / 2] + 1;
		}

		// 3으로 나누어지는 경우 확인
		if (i % 3 == 0 && numbers[i / 3]) {
			numbers[i] = numbers[i] < numbers[i / 3] + 1 ? numbers[i] : numbers[i / 3] + 1;
		}
	}

	printf("%d", numbers[size]);

	return 0;
}

File: 실버 3 (6문제)/1463_1로 만들기.py
# 참고 링크 : https://moon-gd.tistory.com/49
size = int(input())

# size = 1
if size == 1:
    print(0)
# size = 2, 3
elif size <= 3:
    print(1)
# size = 4, 5, 6 ...
else:
    dp = [0] * (size + 1)
    dp[2] = dp[3] = 1
    for i in range(4, size + 1):
        # 우선 i - 1 번째 값 저장 + 1
        dp[i] = dp[i - 1] + 1

        if i % 2 == 0 and dp[i // 2]:
            dp[i] = min(dp[i], dp[i // 2] + 1)

        if i % 3 == 0 and dp[i // 3]:
            dp[i] = min(dp[i], dp[i // 3] + 1)

    print(dp[size])
